### DISCUSSION TITLE
Automatically enable a certain tethering when the software is opened

### DISCUSSION THREAD

**Discussion by SutoLiu** [Mygod/VPNHotspot#563] [Ideas]
I hope to add a custom feature to enable a certain tethering service when the software starts, without having to manually click every time, such as opening the software to automatically enable WiFi hotspots or USB tethering，or Ethernet tethering

**Comment by SutoLiu** [comment]
This can achieve a series of automated Tethering on backup phones, such as when the phone is turned on, the software will automatically start in the background and enable WiFi hotspot sharing or Ethernet Tethering, without the need for manual operation throughout the entire process.
@Mygod

**Reply by sherylynn** [reply]
> This can achieve a series of automated Tethering on backup phones, such as when the phone is turned on, the software will automatically start in the background and enable WiFi hotspot sharing or Ethernet Tethering, without the need for manual operation throughout the entire process. @Mygod

U can try my version, I had make an PR

**Comment by Mygod** [comment]
Why not use Monitor feature + QS tiles?

**Reply by SutoLiu** [reply]
> Why not use Monitor feature + QS tiles?

But when I tried to start the tethering service using the "adb shell cmd statusbar click-tile be.mygod.vpnhotspot/.manage.TetheringTileService$Wifi" command, the WiFi hotspot was not turned on. What is the problem? Can I use commands like "adb shell am start foreground service" to start the hotspot tethering service?
I want to use automation software to automatically start tethering services without having to manually click each time

**Reply by Mygod** [reply]
This might help: https://android.stackexchange.com/a/221910/181631

**Reply by SutoLiu** [reply]
> This might help: https://android.stackexchange.com/a/221910/181631

Thank you very much for your answer.

After adding Ethernet Tethering QS tiles to the status bar and inserting the USB-to-Ethernet adapter, executing the command "cmd statusbar click-tile be.mygod.vpnhotspot/.manage.TetheringTileService$Ethernet" with root privileges did not take effect. Executing the command "cmd statusbar click-tile be.mygod.vpnhotspot/.manage.LocalOnlyHotspotTileService" can be successfully executed, only the QS tiles function under ".manage.TetheringTileService" cannot be enabled by the command.

**Reply by SutoLiu** [reply]
I split the "mobile/src/main/java/be/mygod/vpnhotspot/manage/TetheringTileService.kt" in the source code into four separate Tile Services: "WifiTetheringTileService.kt", "UsbTetheringTileService.kt", "BluetoothTetheringTileService.kt", and "EthernetTetheringTileService.kt". The command can start correctly, perhaps this can be considered a bug

@Mygod

**Reply by SutoLiu** [reply]
It should be that the system executed the content after "$" as a variable, which caused the use of the "cmd statusbar click file" command to be unable to function properly. Splitting it into a separate TileService can be used normally
